pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```diff
@@ -0,0 +1,195 @@
+		return false, errNilMembersCache
+	}
+	defer mc.mux.Unlock()
+
+	mc.mux.Lock()
+	_, ok := mc.cache[key]
+	if !ok {
+		return false, errKeyNotExist
+	}
+	// safe to update
+	mc.cache[key] = *value
+	return true, nil
+}
+
+func (mc *membersCache) delete(key string) (bool, error) {
+	if mc == nil || mc.cache == nil {
+		return false, errNilMembersCache
+	}
+	defer mc.mux.Unlock()
+
+	mc.mux.Lock()
+	_, ok := mc.cache[key]
+	if !ok {
+		return false, errKeyNotExist
+	}
+	// safe to delete
+	delete(mc.cache, key)
+	return true, nil
+}
+
+func (mc *membersCache) get(key string) (etcdserverpb.Member, error) {
+	if mc == nil || mc.cache == nil {
+		return etcdserverpb.Member{}, errNilMembersCache
+	}
+	defer mc.mux.RUnlock()
+
+	mc.mux.RLock()
+	val, ok := mc.cache[key]
+	if !ok {
+		return etcdserverpb.Member{}, errKeyNotExist
+	}
+	return val, nil
+}
+
+// EtcdMemberListerController keeps etcd members cached, and updates the cache when the membership changes
+type EtcdMemberListerController struct {
+	membersCache   *membersCache
+	operatorClient v1helpers.OperatorClient
+	etcdClient     etcdcli.EtcdClient
+}
+
+func NewEtcdMemberListerController(operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &EtcdMemberListerController{
+		operatorClient: operatorClient,
+		etcdClient:     etcdClient,
+		membersCache:   newMembersCache(),
+	}
+	return factory.New().ResyncEvery(30*time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller"))
+}
+
+func (c *EtcdMemberListerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncEtcdMembers(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdMemberListerControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorSyncingEtcdMembers",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "EtcdMemberListerControllerDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "EtcdMembersSynced",
+	}))
+	if updateErr != nil {
+		syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *EtcdMemberListerController) syncEtcdMembers(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+
+	var syncErrs []error
+	// build map from current live members
+	currentMembersMap, errs := c.currentMembersMap(members)
+	if len(errs) > 0 {
+		syncErrs = append(syncErrs, errs...)
```

Comment:
Why are we continuing to check the cache here if we have errors from building the current members map?
Shouldn't we return the errors?

Reply:
well I did not know what to do tbh :) 

So I can proceed with building the map while accumulating errors as I do now, or fail-fast once i encounter an error 

wdyt ? 

cc @tjungblu

Reply:
fail-fast, but only degrade the operator after several consecutive failures.

Reply:
done